src/libraries/Microsoft.PowerFx.Core/Public/Values/DecimalValue.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using System.Text;
+using Microsoft.PowerFx.Core.IR;
+
+namespace Microsoft.PowerFx.Types
+{
+    public class DecimalValue : PrimitiveValue<decimal>
+    {
+        internal DecimalValue(IRContext irContext, decimal value)
+            : base(irContext, value)
+        {
+            bool x = irContext.ResultType == FormulaType.Decimal;
+            if (!x)
+            {
+                x = true;
+            }
+
+            Contract.Assert(IRContext.ResultType == FormulaType.Decimal);
+        }
+
+        public override void Visit(IValueVisitor visitor)
+        {
+            visitor.Visit(this);
+        }
+
+        public decimal Normalize(decimal value)
+        {
+            // remove trailing 0's (significant digits)
+            return Value / 1.000000000000000000000000000000m;
+        }
+
+        public override void ToExpression(StringBuilder sb, FormulaValueSerializerSettings settings)
+        {
+            decimal normalized = Normalize(Value);
```

Comment:
> decimal normalized = Normalize(Value); [](http://example.com/codeflow?start=11&length=39)

The key concern here is round tripping. 
We want to produce an string that when parsed, gives us the value back. 

It's ok to be invariant culture, but I wonder if parse flags could trip us up. 
Is there an unambiguous way to ensure this is a decimal, regardless of parse flag?

Reply:
Funny you should bring this up, I was just looking at the impact of this in src/tests/Microsoft.PowerFx.Interpreter.Tests/PowerFxEvaluationTests.cs as it is has broken some tests.  If the primary mission of ToExpression is to be unambiguous, then yes, for now we can wrap all floating point values in Float( "number" ).  This would be a reason to add a literal suffix, like the `10.2m` or `10.2d` in C#, something I'm not completely opposed to, but was trying to avoid adding that concept, not that I think it is too copmlicated, just more concepts.  :)  Open to your thoughts on this.

Reply:
Actually, I think we should stick to number type invariant.  I'm running all of the expression tests right now in both Decimal and Float modes and with only a few exceptions everything passes.  I think this is a good thing - by default if someone adds a .txt file, it will automatically be tested in both modes.  The exceptions are due to inconsistencies in how C# serializes decimal and float numbers and in numbers that are beyond the decimal range.   I think we should either clean up the serialization inconsistencies and/or walk the tree to do type aware comparisons instead of string comparisons.  We should chat about the approach.